compiler/rustc_hir_analysis/src/collect.rs

```diff
@@ -1388,13 +1463,20 @@ fn infer_return_ty_for_fn_sig<'tcx>(
             let mut recovered_ret_ty = None;
 
             if let Some(suggestable_ret_ty) = ret_ty.make_suggestable(tcx, false, None) {
```

Comment:
> I'm probably missing something obvious... but I'll ask the dumb question anyway:
> 
> Is there a reason we cannot just always use the `keep_erased_ret_ty` here (i.e. always use the ret ty prior to replacing ReErased with ReStatic), rather than doing the work above to compute the value of `keep_erased_lifetime` and then conditionally choosing whether to `ret_ty` or `keep_erased_ret_ty`?
> 
> (I'll give a local build a spin and see if I can find the answer to this myself.)
> 
> (if we _do_ need to keep this logic, then I think I might have an easier time digesting this if the `keep_erased_lifetime` computation were factored out into a separate method rather than being defined inline...)

Thank you.  
Because the changes here only involve a suggestion from E0121.
I did not try understand the subsequent processing process carefully. If keeping the erased lifetime, it will panic in rustc_borrowck.
It seems unreasonable to continue to transfer backwards lifetime `'static` which is not the suggest lifetime `'_`. I think this is indeed an issue worthy of further understanding.

Reply:
> If keeping the erased lifetime, it will panic in rustc_borrowck.

This seems like the important detail here. I had forgotten about this constraint in rustc_borrowck.

Reply:
So, just so I can be concrete about what I am talking about: Here is a diff on your PR that tried out locally:

<details>
<summary>Felix's attempt to eliminate the `keep_erased_lifetime` boolean</summary>

```diff
diff --git a/compiler/rustc_hir_analysis/src/collect.rs b/compiler/rustc_hir_analysis/src/collect.rs
index 19ae890bd1e..c45edba21b1 100644
--- a/compiler/rustc_hir_analysis/src/collect.rs
+++ b/compiler/rustc_hir_analysis/src/collect.rs
@@ -17,7 +17,7 @@
 use rustc_ast::Recovered;
 use rustc_data_structures::captures::Captures;
 use rustc_data_structures::fx::{FxHashSet, FxIndexMap};
-use rustc_data_structures::unord::{UnordMap, UnordSet};
+use rustc_data_structures::unord::UnordMap;
 use rustc_errors::{Applicability, Diag, ErrorGuaranteed, StashKey};
 use rustc_hir as hir;
 use rustc_hir::def::DefKind;
@@ -41,7 +41,6 @@
 use std::cell::Cell;
 use std::iter;
 use std::ops::Bound;
-use std::ops::ControlFlow;

 use crate::check::intrinsic::intrinsic_operation_unsafety;
 use crate::errors;
@@ -1365,7 +1364,7 @@ fn infer_return_ty_for_fn_sig<'tcx>(
     sig: &hir::FnSig<'tcx>,
     generics: &hir::Generics<'_>,
     def_id: LocalDefId,
-    body_id: Option<hir::BodyId>,
+    _body_id: Option<hir::BodyId>,
     icx: &ItemCtxt<'tcx>,
 ) -> ty::PolyFnSig<'tcx> {
     let hir_id = tcx.local_def_id_to_hir_id(def_id);
@@ -1393,68 +1392,6 @@ fn infer_return_ty_for_fn_sig<'tcx>(
             // }
             // -----------------------^--
             //           We should suggest replace `_` with `S<'_>`.
-            let mut keep_erased_lifetime = false;
-            if let Some(body_id) = body_id {
-                struct RetVisitor {
-                    res_hir_ids: UnordSet<hir::HirId>,
-                }
-
-                impl<'v> Visitor<'v> for RetVisitor {
-                    type Result = ControlFlow<()>;
-
-                    fn visit_expr(&mut self, ex: &'v hir::Expr<'v>) -> Self::Result {
-                        if let hir::ExprKind::Path(hir::QPath::Resolved(_, path)) = ex.kind
-                            && let hir::def::Res::Local(hir_id) = path.res
-                            && self.res_hir_ids.contains(&hir_id)
-                        {
-                            ControlFlow::Break(())
-                        } else if let hir::ExprKind::If(_, expr, _) = ex.kind
-                            && let hir::ExprKind::Block(block, _) = expr.kind
-                        {
-                            self.visit_block(block)
-                        } else {
-                            ControlFlow::Continue(())
-                        }
-                    }
-
-                    fn visit_block(&mut self, b: &'v hir::Block<'v>) -> Self::Result {
-                        if let Some(ret) = b.expr {
-                            self.visit_expr(ret)
-                        } else if let Some(ret) = b.stmts.last() {
-                            self.visit_stmt(ret)
-                        } else {
-                            ControlFlow::Continue(())
-                        }
-                    }
-
-                    fn visit_stmt(&mut self, s: &'v hir::Stmt<'v>) -> Self::Result {
-                        if let hir::StmtKind::Semi(expr) = s.kind
-                            && let hir::ExprKind::Ret(Some(ret)) = expr.kind
-                        {
-                            self.visit_expr(ret)
-                        } else {
-                            ControlFlow::Continue(())
-                        }
-                    }
-
-                    fn visit_item(&mut self, _i: &'v hir::Item<'v>) -> Self::Result {
-                        ControlFlow::Continue(())
-                    }
-                }
-
-                let body = tcx.hir().body(body_id);
-                if let hir::ExprKind::Block(b, _) = body.value.kind {
-                    let mut res_hir_ids = UnordSet::new();
-                    for param in body.params {
-                        res_hir_ids.insert(param.pat.hir_id);
-                    }
-                    let mut ret_visitor = RetVisitor { res_hir_ids };
-                    if let ControlFlow::Break(()) = ret_visitor.visit_block(b) {
-                        keep_erased_lifetime = true;
-                    }
-                }
-            }
-
             let mut diag = bad_placeholder(tcx, visitor.0, "return type");
             let ret_ty = fn_sig.output();
             // Don't leak types into signatures unless they're nameable!
@@ -1468,9 +1405,7 @@ fn visit_item(&mut self, _i: &'v hir::Item<'v>) -> Self::Result {
                 diag.span_suggestion(
                     ty.span,
                     "replace with the correct return type",
-                    if keep_erased_lifetime
-                        && let Some(ty) = keep_erased_ret_ty.make_suggestable(tcx, false, None)
-                    {
+                    if let Some(ty) = keep_erased_ret_ty.make_suggestable(tcx, false, None) {
                         ty
                     } else {
                         suggestable_ret_ty
```

</summary>

WIth that diff in place, I didn't see any new borrow check failures (I think).

Here's what I *did* see, in terms of changes to the UI tests:

    [ui] tests/ui/suggestions/return-cycle-2.rs
    [ui] tests/ui/typeck/typeck_type_placeholder_item.rs

with these associated deltas to the UI output with my patch applied:

```
---- [ui] tests/ui/suggestions/return-cycle-2.rs stdout ----
diff of stderr:

5          |                                  ^
6          |                                  |
7          |                                  not allowed in type signatures
-          |                                  help: replace with the correct return type: `Token<&'static T>`
+          |                                  help: replace with the correct return type: `Token<&T>`
9
10      error: aborting due to 1 previous error
11
```

and 

```
---- [ui] tests/ui/typeck/typeck_type_placeholder_item.rs stdout ----
diff of stderr:

158        |                         -^
159        |                         ||
160        |                         |not allowed in type signatures
-          |                         help: replace with the correct return type: `&'static &'static usize`
+          |                         help: replace with the correct return type: `&&usize`
162
163     error[E0121]: the placeholder `_` is not allowed within types on item signatures for return types
164       --> $DIR/typeck_type_placeholder_item.rs:53:52

562        |               ----------------^-
563        |               |               |
564        |               |               not allowed in type signatures
-          |               help: replace with the correct return type: `Option<&'static u8>`
+          |               help: replace with the correct return type: `Option<&u8>`
566
567     error[E0121]: the placeholder `_` is not allowed within types on item signatures for constants
568       --> $DIR/typeck_type_placeholder_item.rs:222:10
```

So, my question: which is right? What *should* these two tests be suggesting here?

Reply:
(Also, in case its not clear: I'm not suggesting you just blindly adopt the diff I posted in the previous comment. It was meant as a sketch as, I am trying to understand the significance of the different control flow branches in your PR, and not as an idealized form of the code.)

Reply:
(a further note: If I go one step further with the diff I posted above, and attempt to actually *use* the result of `keep_erased_ret_ty.make_suggestable(tcx, false, None)` as the new value that we write into `recovered_ret_ty`, *then* I *do* see an ICE during borrowck saying "cannot convert `'{erased}` to a region vid". So yes, I do entirely believe that borrowck failures are witnessable here. I'm just trying to make sure we don't overindex on that point....)